packages/ddex/ingester/indexer/indexer.go

```diff
@@ -1,62 +1,266 @@
+		return
+	}
+
+	// Unzip the file and process its contents
+	extractDir, err := os.MkdirTemp("", "extracted")
+	if err != nil {
+		i.logger.Error("Error creating temp directory", "error", err)
+		return
+	}
+	defer os.RemoveAll(extractDir)
+
+	if err := unzip(zipFilePath, extractDir); err != nil {
+		i.logger.Error("Error unzipping file", "error", err)
+		return
+	}
 
-	delete(fullDocument, "_id")
+	i.processZIPContents(extractDir)
+}
 
-	// TODO process upload:
-	// 1. download zip from raw s3 bucket
-	// 2. unzip
-	// 3. upload files to indexed s3 bucket
+// processZIPContents finds deliveries in rootDir (and its subdirectories), uploads their contents to S3, and inserts them into the MongoDB "indexed" collection.
+func (i *Indexer) processZIPContents(rootDir string) error {
+	// The root directory could be the single delivery
+	i.processDelivery(rootDir, rootDir)
 
-	// Write delivery to 'indexed' collection
-	fullDocument["delivery_status"] = constants.DeliveryStatusValidating
-	// TODO download xml from bucket
-	// fullDocument["delivery_xml"] = ...
+	// Or it could contain multiple deliveries
+	return filepath.Walk(rootDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if strings.Contains(path, "__MACOSX") {
+			return nil
+		}
+		if info.IsDir() {
+			if path == rootDir {
+				return nil
+			}
+			return i.processDelivery(rootDir, path)
+		}
+		return nil
+	})
+}
 
-	result, err := indexedColl.InsertOne(ctx, fullDocument)
+// processDelivery parses a "delivery" from dir if dir contains an XML file.
+func (i *Indexer) processDelivery(rootDir, dir string) error {
+	var deliveryID, xmlRelativePath string
+	var xmlBytes []byte
+	files, err := os.ReadDir(dir)
 	if err != nil {
-		log.Fatal(err)
+		return fmt.Errorf("failed to read directory %s: %w", dir, err)
+	}
+
+	// Look for the XML first and make a new deliveryID for it
+	for _, file := range files {
+		if !file.IsDir() && strings.HasSuffix(file.Name(), ".xml") {
+			deliveryID = ulid.Make().String()
+			xmlPath := filepath.Join(dir, file.Name())
+			xmlBytes, err = os.ReadFile(xmlPath)
+			if err != nil {
+				return fmt.Errorf("failed to read XML file %s: %w", xmlPath, err)
+			}
+
+			xmlRelativePath, err = filepath.Rel(rootDir, xmlPath)
+			if err != nil {
+				return fmt.Errorf("failed to compute relative path for %s: %w", xmlPath, err)
+			}
+			break // Assume only one XML file per directory
+		}
+	}
+
+	if deliveryID == "" {
+		i.logger.Info("No XML file found in directory. Skipping", "dir", dir)
+		return nil
+	}
+
+	// Upload the delivery's audio and images to S3
+	for _, file := range files {
+		if !file.IsDir() && !strings.HasSuffix(file.Name(), ".xml") {
+			filePath := filepath.Join(dir, file.Name())
+			i.uploadToS3Indexed(filePath, fmt.Sprintf("%s/%s", deliveryID, file.Name()))
+		}
+	}
+
+	// Insert the delivery into the Mongo "indexed" collection
+	deliveryDoc := bson.M{
```

Comment:
should this also contain a reference to the original upload in the `uploads` collection? e.g the upload etag or _id

also, since we've refined the terminology and 1 upload has many "deliveries", can we rename `delivery_etag` in `uploads` to `upload_etag`

Reply:
for sure, just made those changes now